lib/PipelineC/PipelineCTracing.cpp

```diff
@@ -0,0 +1,283 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <mutex>
+#include <optional>
+#include <thread>
+
+#include "llvm/Support/Base64.h"
+#include "llvm/Support/Process.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "revng/ADT/ConstexprString.h"
+#include "revng/PipelineC/PipelineC.h"
+#include "revng/PipelineC/TracingCommon.h"
+#include "revng/Support/Assert.h"
+
+inline constexpr auto TracingEnv = "REVNG_C_API_TRACE_PATH";
+static bool TracingEnabled = llvm::sys::Process::GetEnv(TracingEnv).hasValue();
+
+// The opposite of a std::recursive_mutex, if locked by the same thread it will
+// assert (this is to avoid a deadlock/malformed output when tracing)
+class NonRecursiveMutex {
+private:
+  std::optional<std::thread::id> ThreadId;
+  std::mutex TheLock;
+
+public:
+  NonRecursiveMutex() {}
+
+  void lock() {
+    if (ThreadId.has_value())
+      revng_assert(std::this_thread::get_id() != *ThreadId,
+                   "NonRecursiveMutex entered twice by the same thread!");
+    TheLock.lock();
+    ThreadId = std::this_thread::get_id();
+  }
+
+  void unlock() {
+    revng_assert(std::this_thread::get_id() == *ThreadId);
+    TheLock.unlock();
+    ThreadId.reset();
+  }
+};
+
+static NonRecursiveMutex TraceLock;
+
+// Helper class for tracing, this will be used by the the argument/return
+// handlers defined below to properly print the value of the arguments onto the
+// YAML tracing file
+class PipelineCTracer {
+private:
+  std::string Path;
+  std::error_code OpenEC;
+  llvm::raw_fd_ostream OS;
+  bool OutputtingArguments;
+
+public:
+  PipelineCTracer(std::string Path) :
+    Path(Path), OpenEC(), OS(Path, OpenEC), OutputtingArguments(false) {
+    revng_assert(!OpenEC);
+    printHeader();
+  }
+
+  ~PipelineCTracer() {
+    OS.flush();
+    OS.close();
+  }
+
+private:
+  void printHeader() {
+    OS << "version: 1\n";
+    OS << "commands:\n";
+    OS.flush();
+  }
+
+  std::string reprString(const char *String) {
+    return '"' + llvm::StringRef(String).str() + '"';
```

Comment:
1. No escape?
2. I think `std::string` has a constructor for this, no need to go through `StringRef`

Reply:
1. Done
2. Removed cast